test/shape_transform_descriptor.cpp

```diff
@@ -0,0 +1,389 @@
+                         make_op("reshape", {{"dims", {3, 8, 8}}}),
+                     });
+}
+
+TEST_CASE(optimize_resize2)
+{
+    EXPECT(migraphx::optimize_shape_transforms(
+               {1, 1, 2, 2},
+               {
+                   make_op("reshape", {{"dims", {1, 1, 2, 1, 2, 1}}}),
+                   make_op("multibroadcast", {{"out_lens", {1, 2, 2, 2, 2, 3}}}),
+                   make_op("reshape", {{"dims", {1, 2, 4, 6}}}),
+               }) == ops{
+                         make_op("unsqueeze", {{"axes", {3, 5}}}),
+                         make_op("multibroadcast", {{"out_lens", {1, 1, 2, 2, 2, 3}}}),
+                         make_op("reshape", {{"dims", {1, 1, 4, 6}}}),
+                         make_op("multibroadcast", {{"out_lens", {1, 2, 4, 6}}}),
+                     });
+}
+
+TEST_CASE(optimize_reshape_2_squeeze)
+{
+    EXPECT(migraphx::optimize_shape_transforms({3, 1, 5, 1, 2, 1, 1},
+                                               {
+                                                   make_op("reshape", {{"dims", {3, 5, 2}}}),
+                                               }) ==
+           ops{
+               make_op("squeeze", {{"axes", {1, 3, 5, 6}}}),
+           });
+}
+
+TEST_CASE(optimize_reshape_2_unsqueeze)
+{
+    EXPECT(migraphx::optimize_shape_transforms(
+               {3, 5, 2},
+               {
+                   make_op("reshape", {{"dims", {3, 1, 5, 1, 2, 1, 1}}}),
+               }) == ops{
+                         make_op("unsqueeze", {{"axes", {1, 3, 5, 6}}}),
+                     });
+}
+
+TEST_CASE(optimize_unsqueeze_multibroadcast)
+{
+    EXPECT(migraphx::optimize_shape_transforms(
+               {32, 10},
+               {
+                   make_op("unsqueeze", {{"axes", {0, 3, 4}}}),
+                   make_op("multibroadcast", {{"out_lens", {256, 32, 10, 16, 16}}}),
+               }) == ops{
+                         make_op("broadcast", {{"axis", 1}, {"out_lens", {256, 32, 10, 16, 16}}}),
+                     });
+}
+
+TEST_CASE(optimize_multibroadcast_reshape)
+{
+    EXPECT(migraphx::optimize_shape_transforms(
+               {1, 4, 1},
+               {
+                   make_op("multibroadcast", {{"out_lens", {2, 4, 6}}}),
+                   make_op("reshape", {{"dims", {2, 2, 2, 6}}}),
+               }) == ops{
+                         make_op("reshape", {{"dims", {1, 2, 2, 1}}}),
+                         make_op("multibroadcast", {{"out_lens", {2, 2, 2, 6}}}),
+                     });
+}
+
+TEST_CASE(optimize_squeeze_broadcast)
+{
+    EXPECT(migraphx::optimize_shape_transforms(
+               {256, 1, 1},
+               {
+                   make_op("squeeze"),
+                   make_op("broadcast", {{"axis", 0}, {"out_lens", {256, 64, 1, 1}}}),
+               }) == ops{
+                         make_op("unsqueeze", {{"axes", {3}}}),
+                         make_op("multibroadcast", {{"out_lens", {256, 64, 1, 1}}}),
```

Comment:
It looks like this code will prioritize `unsqueeze` over `squeeze`? Is there a benefit to this or is it arbitrary?

Reply:
I dont know if its prioritization, but rather the way we interpret transforms.

This starts with `[[256:0]], [[64:]], [[1:1]], [[1:2]]`, so it first tries to insert the broadcast. The `broadcast` op wont work for this case since the "succeeding sizes must match"(which it would be nice to remove this restriction in the future), so it inserts a multibroadcast instead. 

Then it flattens it to `[[256:0]], [[1:]], [[1:1]], [[1:2]]`, which will insert an unsqueeze to create the dimension which has an empty axis.